### PR TITLE
HttpClient leaves connection open on websocket request failure (resource leak) - fixes #1757

### LINKED ISSUE #1757: HttpClient leaves connection open on websocket request failure (resource leak)
When running this code:
```java
        Vertx vertx = Vertx.vertx();
        vertx.createHttpServer().requestHandler(req -> {
            req.response().setStatusCode(200);
            req.response().headers().set("Connection", "close");
            req.response().end();
        }).listen(44444);
        vertx.createHttpClient().websocket(44444, "localhost", "/", ws -> {
            System.out.println("WS Unexpectedly connected");
        }, t -> {
            System.err.println("WS Connection failed");
            t.printStackTrace();
        });
```
Then end result is `WS Connection failed` (as expected) but `netstat -natp|grep :44444` (in Linux) shows the connection is still ESTABLISHED. Only closing the java process (naturally) gets rid of the connection.

For a regular GET request:
```java
        Vertx vertx = Vertx.vertx();
        vertx.createHttpServer().requestHandler(req -> {
            req.response().setStatusCode(200);
            req.response().headers().set("Connection", "close");
            req.response().end();
        }).listen(44444);
        vertx.createHttpClient().get(44444, "localhost", "/", c -> {
            System.out.println("GET connected");
        }).end();
```

then `netstat -natp|grep :44444` shows no ESTABLISHED connections, as expected.

I added the `Connection: close` header so that HttpClient wouldn't pool the connection and leave it open because of that.

### PR DESCRIPTION
**Pull request by xkr47**
Continuing from #1757:

Ok, I tried, but also had to add `handshaker = null;` so that it wouldn't try to send a graceful websocket frame to shut it down.

This seems to work fine for me with both versions of the code posted in #1757. However I do not know if there are further resource cleanups that need to be done, maybe related to connection pools, so please consider that.

### CODE REVIEW

**Comment by vietj** [comment]
can you provide the corresponding tests ?

**Comment by xkr47** [comment]
I'll try..

**Comment by xkr47** [comment]
Ok, hope this is good enough...

File: src/main/java/io/vertx/core/http/impl/ClientConnection.java [hunk 1]
```diff
@@ -189,6 +189,8 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
         if (msg instanceof HttpResponse) {
           HttpResponse resp = (HttpResponse) msg;
           if (resp.getStatus().code() != 101) {
+            handshaker = null;
+            close();
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: I think close should be called after the handleException statement

File: src/test/java/io/vertx/test/core/Issue1757_WebsocketFailureTest.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                        ws.close();
+                    }, t -> {
+                        System.out.println("Client got exception");
+                        addResult(t);
+                    });
+                });
+            }
+        });
+        try {
+            System.out.println("Test thread waiting for resultQueue");
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/Issue1757_WebsocketFailureTest.java]: no System.out

File: src/main/java/io/vertx/core/http/impl/ClientConnection.java (hunk 1 above)
Comment (xkr47) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: I just copypasted from line 215-218.. I have no clue which way is better..

File: src/test/java/io/vertx/test/core/Issue1757_WebsocketFailureTest.java (hunk 2 above)
Comment (xkr47) [inline src/test/java/io/vertx/test/core/Issue1757_WebsocketFailureTest.java]: All System.out removed.

File: src/main/java/io/vertx/core/http/impl/ClientConnection.java (hunk 1 above)
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: ok

**Comment by vietj** [comment]
can you move the test to the `WebsocketTest` class ?

File: src/test/java/io/vertx/test/core/WebsocketTest.java [hunk 3]
```diff
@@ -1350,4 +1357,86 @@ public void testClientWebsocketWithHttp2Client() throws Exception {
     }));
     await();
   }
+
+  @Test
+  public void testIssue1757_1() throws Throwable {
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/WebsocketTest.java]: can you rename this to something more meaningful ?

File: src/test/java/io/vertx/test/core/WebsocketTest.java [hunk 4]
```diff
@@ -1350,4 +1357,86 @@ public void testClientWebsocketWithHttp2Client() throws Exception {
...
+    doTestIssue1757(true);
+  }
+
+  @Test
+  public void testIssue1757_2() throws Throwable {
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/WebsocketTest.java]: can you rename this to something more meaningful ?

File: src/test/java/io/vertx/test/core/WebsocketTest.java (hunk 3 above)
Comment (vietj) [inline src/test/java/io/vertx/test/core/WebsocketTest.java]: @xkr47 ?

File: src/test/java/io/vertx/test/core/WebsocketTest.java (hunk 4 above)
Comment (vietj) [inline src/test/java/io/vertx/test/core/WebsocketTest.java]: @xkr47 ?

**Comment by vietj** [comment]
thanks for the contrib!

**Pull request merged**